cygwin/CygwinMachine.c

```diff
@@ -0,0 +1,327 @@
+/*
+htop - CygwinMachine.c
+(C) 2014 Hisham H. Muhammad
+(C) 2015 Michael McConville
+(C) 2017,2018 Guy M. Broome
+Released under the GNU GPLv2+, see the COPYING file
+in the source distribution for its full text.
+*/
+
+#include "cygwin/CygwinMachine.h"
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include "CRT.h"
+#include "Settings.h"
+#include "XUtils.h"
+
+
+static void CygwinMachine_updateCPUcount(CygwinMachine* this) {
+   Machine* super = &this->super;
+   long int s;
+
+   s = sysconf(_SC_NPROCESSORS_CONF);
+   if (s < 1)
+      CRT_fatalError("Cannot get existing CPU count by sysconf(_SC_NPROCESSORS_CONF)");
+
+   if (s != super->existingCPUs) {
+      if (s == 1) {
+         this->cpuData = xRealloc(this->cpuData, sizeof(CPUData));
+         this->cpuData[0].online = true;
+      } else {
+         this->cpuData = xReallocArray(this->cpuData, s + 1, sizeof(CPUData));
+         this->cpuData[0].online = true; /* average is always "online" */
+         for (int i = 1; i < s + 1; i++) {
+            this->cpuData[i].online = true;  // TODO: support offline CPUs and hot swapping
+         }
+      }
+
+      super->existingCPUs = s;
+   }
+
+   s = sysconf(_SC_NPROCESSORS_ONLN);
+   if (s < 1)
+      CRT_fatalError("Cannot get active CPU count by sysconf(_SC_NPROCESSORS_ONLN)");
+
+   if (s != super->activeCPUs) {
+      super->activeCPUs = s;
+   }
+}
+
+Machine* Machine_new(UsersTable* usersTable, uid_t userId) {
+   CygwinMachine* this = xCalloc(1, sizeof(CygwinMachine));
+   Machine* super = &this->super;
+
+   Machine_init(super, usersTable, userId);
+
+   // Initialize page size
+   if ((this->pageSize = sysconf(_SC_PAGESIZE)) == -1)
+      CRT_fatalError("Cannot get pagesize by sysconf(_SC_PAGESIZE)");
+   this->pageSizeKB = this->pageSize / ONE_K;
+
+   // Initialize clock ticks
+   if ((this->jiffies = sysconf(_SC_CLK_TCK)) == -1)
+      CRT_fatalError("Cannot get clock ticks by sysconf(_SC_CLK_TCK)");
+
+   // Read btime (the kernel boot time, as number of seconds since the epoch)
+   FILE* statfile = fopen(PROCSTATFILE, "r");
+   if (statfile == NULL)
+      CRT_fatalError("Cannot open " PROCSTATFILE);
+
+   this->boottime = -1;
+
+   while (true) {
+      char buffer[PROC_LINE_LENGTH + 1];
+      if (fgets(buffer, sizeof(buffer), statfile) == NULL)
+         break;
+      if (String_startsWith(buffer, "btime ") == false)
+         continue;
+      if (sscanf(buffer, "btime %lld\n", &this->boottime) == 1)
+         break;
+      CRT_fatalError("Failed to parse btime from " PROCSTATFILE);
+   }
+   fclose(statfile);
+
+   if (this->boottime == -1)
+      CRT_fatalError("No btime in " PROCSTATFILE);
```

Comment:
Using [`GetTickCount64`](https://learn.microsoft.com/de-de/windows/win32/api/sysinfoapi/nf-sysinfoapi-gettickcount64?redirectedfrom=MSDN) in combination with [`clock_gettime`](https://linux.die.net/man/3/clock_gettime) and `CLOCK_REALTIME` is both faster, less code, and avoids most of the error checking.